In [ ]:
# Point Colaboratory to your Google Drive

from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
datapath = '/content/gdrive/My Drive/^DATA SCIENCE/^Biological Data/PROJECT/models/' # Luca
# datapath = '/content/gdrive/My Drive/Data Science - Unipd/II Year/I Semester/Biological Data/Project/PROJECT/models/' # Giuliano

# Building models

## Ex 5: build a PSSM model

In [ ]:
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_msa ./PROJECT/clustal_omega_uniprot.fasta -db ./data/msa/uniprot_sprot.fasta -out ./PROJECT/output_psiblast_uniprot.out -out_pssm ./PROJECT/pssm_model_uniprot.pssm
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_msa ./PROJECT/clustal_omega_uniref50.fasta -db ./data/msa/uniprot_sprot.fasta -out ./PROJECT/output_psiblast_uniref50.out -out_pssm ./PROJECT/pssm_model_uniref50.pssm
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_msa ./PROJECT/clustal_omega_uniref90.fasta -db ./data/msa/uniprot_sprot.fasta -out ./PROJECT/output_psiblast_uniref90.out -out_pssm ./PROJECT/pssm_model_uniref90.pssm

## Ex 6: build a HMM model

In [ ]:
# ./binx/hmmer-3.3.1/src/hmmbuild ./PROJECT/hmm_model_uniprot.hmm ./PROJECT/clustal_omega_uniprot.fasta
# ./binx/hmmer-3.3.1/src/hmmbuild ./PROJECT/hmm_model_uniref50.hmm ./PROJECT/clustal_omega_uniref50.fasta
# ./binx/hmmer-3.3.1/src/hmmbuild ./PROJECT/hmm_model_uniref90.hmm ./PROJECT/clustal_omega_uniref90.fasta

## Ex 7: find significant hits using HMM-SEARCH and PSI-BLAST

In [ ]:
# hmmsearch --tblout ./PROJECT/hmm_search_uniprot.hmmer_tblout --domtblout ./PROJECT/hmm_search_uniprot.hmmer_domtblout ./PROJECT/hmm_model_uniprot.hmm ./data/msa/uniprot_sprot.fasta > ./PROJECT/hmm_search_uniprot.hmmer_align
# hmmsearch --tblout ./PROJECT/hmm_search_uniref50.hmmer_tblout --domtblout ./PROJECT/hmm_search_uniref50.hmmer_domtblout ./PROJECT/hmm_model_uniref50.hmm ./data/msa/uniprot_sprot.fasta > ./PROJECT/hmm_search_uniref50.hmmer_align
# hmmsearch --tblout ./PROJECT/hmm_search_uniref90.hmmer_tblout --domtblout ./PROJECT/hmm_search_uniref90.hmmer_domtblout ./PROJECT/hmm_model_uniref90.hmm ./data/msa/uniprot_sprot.fasta > ./PROJECT/hmm_search_uniref90.hmmer_align

In [ ]:
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_pssm ./PROJECT/pssm_model_uniprot.pssm -db ./data/msa/uniprot_sprot.fasta -outfmt 6 -num_iterations 4 -evalue 0.01 > ./PROJECT/psiblast_search_uniprot.txt
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_pssm ./PROJECT/pssm_model_uniref50.pssm -db ./data/msa/uniprot_sprot.fasta -outfmt 6 -num_iterations 4 -evalue 0.01 > ./PROJECT/psiblast_search_uniref50.txt
# ./binx/ncbi-blast-2.11.0+/bin/psiblast -in_pssm ./PROJECT/pssm_model_uniref90.pssm -db ./data/msa/uniprot_sprot.fasta -outfmt 6 -num_iterations 4 -evalue 0.01 > ./PROJECT/psiblast_search_uniref90.txt

## Ex 8

#### Read the ground truth file


In [ ]:
!pip install Bio
from Bio import SeqIO

     |████████████████████████████████| 71kB 3.6MB/s 
     |████████████████████████████████| 2.3MB 6.3MB/s 


Find the proteins containing the target domain (PF08543) in:
- SwissProt (annotated) 
- TrEMBLE (not annotated) ?????

Find the proteins containing the target domain (PF08543) in SwissProt -> set of actual positives 


In [ ]:
# !pip install Bio
# from Bio import SeqIO
path = datapath + 'protein-matching-PF08543.fasta'
protein_matching = SeqIO.parse(path,'fasta')

list_protein_containing_domain = []
for sequence in protein_matching:
    # print(sequence)
    name = sequence.id # name is in the form A1ADT5|reviewed|Pyridoxine/pyridoxal/pyridoxamine
    # print('name ->', name, '\n')
    list_protein_containing_domain.append(name.split('|')[0])

set_protein_containing_domain = set(list_protein_containing_domain) # this is the ground truth
print(f"Number of proteins in SwissProt containing the target domain: {len(list_protein_containing_domain)}")
# this is the total number of protein in SwissProt
print(f"Actual positives: {set_protein_containing_domain}")

Number of proteins in SwissProt containing the target domain: 150
Actual positives: {'Q8W1X2', 'Q62LP6', 'P99124', 'B7M6S8', 'Q2L1P5', 'Q5PNC8', 'Q63SC2', 'A3N2D3', 'Q8CTQ7', 'Q5PIK8', 'Q8Z4W1', 'P9WG77', 'Q7WII1', 'Q9RYX0', 'Q6GEY2', 'P76422', 'Q7N3W7', 'B7V753', 'Q1I2L8', 'Q0BSF0', 'Q1J237', 'B2JCI0', 'Q1AYE5', 'Q6D5V1', 'Q9ZL00', 'P56904', 'Q57LS3', 'Q8NQH1', 'Q39I40', 'Q8D4Q2', 'Q8NVH3', 'A6VEZ4', 'Q8K183', 'P40191', 'Q51892', 'B0UUD2', 'O35331', 'Q83KY1', 'B7MY71', 'Q57688', 'Q1CIM6', 'P55882', 'Q1R8V2', 'Q48BL6', 'Q9CNY1', 'Q8ZPM8', 'Q7MGA4', 'A1ADT5', 'Q87FP6', 'P66915', 'Q0II59', 'Q8FTH8', 'Q320Z3', 'P44697', 'O94265', 'A5UA83', 'B1JFM7', 'Q57PI7', 'B7LCG3', 'Q3K4B8', 'Q52975', 'B0KR83', 'O25515', 'O74860', 'B1IX53', 'Q5HRG7', 'Q8Z6Q3', 'Q8FFB5', 'P44690', 'Q0TF48', 'O83153', 'O85786', 'O94266', 'Q141E8', 'Q06490', 'B7UGB9', 'B7LL66', 'Q0THJ1', 'O31620', 'Q5M731', 'B7MHS2', 'Q6G7L7', 'Q2SXQ4', 'Q88C26', 'Q1RBF9', 'P40192', 'B1XA89', 'O67772', 'P66914', 'B5YZW5', 'Q55EK9', 'Q6LP

Find all the proteins in the target database (uniprot_sprot.fasta) -> complete ground truth (actual positives + actual negatives)

In [ ]:
path = datapath + 'uniprot_sprot.fasta'
all_sequences = SeqIO.parse(path, 'fasta')

list_all_sequences = []
for sequence in all_sequences:
    name = sequence.id # name is in the form sp|Q6GZX4|001R_FRG3G
    list_all_sequences.append(name.split('|')[1])

num_sequences = len(set(list_all_sequences))
print(f"Total number of proteins in SwissProt: {num_sequences}")

Total number of proteins in SwissProt: 563972


#### Read the output file of the psi-blast search

In [ ]:
def read_psiblast(path):
    ''' Create a dictionary having as key the protein ID and as value the set of
    tuples (start, end) representing the position of our domain in that protein
    (we assume that the domain may be present in more than one position). '''
    
    positions = {}
    with open(path, 'r') as f:
        for line in f:
            if line:
                line = line.replace('|', ' ')
                (_, _, id, _, _, length, _, gapopen, _, _, 
                    start, end, evalue, bitscore) = line.split()
                    
                positions.setdefault(id, set()).add((start, end)) 
    return positions

In [ ]:
path = datapath + 'psiblast_search_uniref90.txt'
positions_psiblast = read_psiblast(path)
print(f'Number of distint proteins found with psi-blast: {len(positions_psiblast)}')

Number of distint proteins found with psi-blast: 674


In [ ]:
positions_psiblast

{'A0A0H2WZY4': {('162', '278')},
 'A0AFC3': {('15', '247'), ('20', '244'), ('21', '244')},
 'A0KPL4': {('124', '317'), ('189', '287')},
 'A0KTG8': {('124', '295'), ('189', '287')},
 'A0LP26': {('12', '247'), ('22', '244')},
 'A0Q299': {('12', '240'), ('7', '246'), ('83', '237')},
 'A0R980': {('13', '249'), ('27', '249'), ('63', '240')},
 'A1A2J2': {('23', '276')},
 'A1A6H3': {('243', '357')},
 'A1ACX0': {('22', '255'), ('25', '252')},
 'A1ADT5': {('14', '256'), ('14', '269'), ('18', '272')},
 'A1AFX4': {('124', '287'), ('124', '295')},
 'A1AV12': {('73', '298'), ('73', '308')},
 'A1JQJ2': {('24', '250'), ('3', '250'), ('45', '225')},
 'A1JQV6': {('121', '295')},
 'A1RGD1': {('124', '295'), ('189', '287')},
 'A1S0R2': {('321', '476')},
 'A1SX95': {('12', '257')},
 'A1SZY2': {('124', '296'), ('190', '288')},
 'A1TYR4': {('191', '285')},
 'A1VBV4': {('10', '249')},
 'A2BLC0': {('248', '488')},
 'A2RGF2': {('179', '276')},
 'A2RKJ5': {('18', '240'), ('3', '240')},
 'A2SU11': {('22', '245')

#### Read the output file of the hmm search

In [ ]:
def read_hmm(path):

    ''' Create a dictionary having as key the protein ID and as value the set of
    tuples (start, end) representing the position of our domain in that protein
    (we assume that the domain may be present in more than one position). '''
  
    positions = {}
    with open(path,'r') as f:
        for line in f:
            if line[0] != "#": # and line[0] != "-":
              line=line.replace('|', ' ')
              (_, id, _, _, tlen, _, _, qlen,  
              full_seq_evalue, full_seq_score, _, _, _, dcevalue, dievalue, dscore, _, _, _ , 
              align_from, align_to, env_from, env_to, accuracy, description) = line.strip().split()[:25]
              #exist multiple scores and alignment. Above I took all of them, but in the below dictionary
              #I have put the same used in github reference
            
              positions.setdefault(id, set()).add( (align_from, align_to) )
   
    return positions


In [ ]:
path = datapath + 'hmm_search_uniref90.hmmer_domtblout'
positions_hmm = read_hmm(path)
print(f'Number of distint proteins found with hmm: {len(positions_hmm)}')

Number of distint proteins found with hmm: 412


In [ ]:
positions_hmm

{'A0A0H2WZY4': {('167', '228')},
 'A0AFC3': {('47', '219')},
 'A0KPL4': {('189', '288')},
 'A0KTG8': {('189', '288')},
 'A0R980': {('189', '238')},
 'A1A6H3': {('241', '348')},
 'A1ACX0': {('121', '241')},
 'A1ADT5': {('88', '276')},
 'A1AFX4': {('189', '287')},
 'A1AV12': {('199', '299')},
 'A1JQJ2': {('67', '226')},
 'A1RGD1': {('189', '288')},
 'A1S3I3': {('189', '289')},
 'A1SZY2': {('190', '289')},
 'A3D837': {('189', '288')},
 'A3MZC0': {('189', '288')},
 'A3N2D3': {('73', '270')},
 'A3NJS6': {('93', '244')},
 'A3P5E2': {('93', '244')},
 'A3QAN3': {('189', '288')},
 'A4SIG6': {('189', '288')},
 'A4VR47': {('188', '287')},
 'A4WEI4': {('189', '288')},
 'A4XPU9': {('188', '287')},
 'A4Y9Z9': {('189', '288')},
 'A4YYB6': {('198', '303')},
 'A5EN78': {('198', '302')},
 'A5EWS4': {('190', '288')},
 'A5G6F4': {('199', '299')},
 'A5HYZ4': {('62', '206')},
 'A5N1D3': {('67', '210')},
 'A5UA83': {('74', '273')},
 'A5UCC3': {('189', '287')},
 'A5UEH8': {('189', '287')},
 'A5W9W6': {('188',

#### Model evaluation

In [ ]:
# !pip install Bio
# from Bio import SeqIO
import math
import json
import pandas as pd
import numpy as np

#### Useful functions to compute metrics

CONFIDENTIAL!!! :-)

In $\verb!utils.py!$ there are 5 functions:

- $\verb!computeMetrics!$
- $\verb!createPositionSet!$
- $\verb!evaluateSequencesSH2!$
- $\verb!getSeqLength!$
- $\verb!evaluatePositionsSH2!$

But for point 8) are needed only:
- $\verb!computeMetrics!$
- $\verb!evaluateSequencesSH2!$

So I use only them in the following.

In [ ]:
# import math
def metrics(true_positive, true_negative, false_positive, false_negative):
  
    accuracy = (true_positive + true_negative) / (true_positive + true_negative + false_negative + false_positive)
    precision = true_positive / (true_positive + false_positive)
    sensitivity = true_positive / (true_positive + false_negative)
    specificity = true_negative / (true_negative + false_positive)
    weighted_accuracy = (sensitivity + specificity) / 2
    mcc_numerator = (true_positive * true_negative) - (false_positive * false_negative)
    mcc_denominator_squared = (true_positive + false_positive) * (true_positive + false_negative) * (true_negative + false_positive) * (true_negative + false_negative)
    mcc = mcc_numerator / math.sqrt(mcc_denominator_squared)
    f1_score = 2 * (precision * sensitivity) / (precision + sensitivity)
    
    return {
            "Accuracy": accuracy,
            "Weighted accuracy": weighted_accuracy,
            "Precision": precision,
            "Sensitivity": sensitivity,
            "Specificity": specificity,
            "MCC": mcc,
            "F-score": f1_score
            }


In [ ]:
def metrics_results(predictions, actual_positives, ground_truth_size):
    """
    predictions = dict of the type {'A0A0H2WZY4': {('167', '228')}, ... }
    actual_positives = set of sequences containing PF08543
    ground_truth_size = number of actual positives + actual negatives (= SwissPort size)
    """
    predicted_sequences = set(predictions.keys())
    
    true_positive = len(predicted_sequences & actual_positives)
    
    # # If there aren't true postive later there will be a division by 0
    # if true_positive == 0:
    #     print("No true positive!")
    #     return
    
    false_negative = len(actual_positives) - true_positive
    false_positive = len(predicted_sequences) - true_positive
    true_negative = ground_truth_size - true_positive - false_positive - false_negative
    
    conf_matrix = pd.DataFrame(np.array([[true_positive, false_positive], [false_negative, true_negative]]), 
                               index=['Positive pred', 'Negative pred'], 
                               columns=['Actual positive', 'Actual negative'])

    return conf_matrix, metrics(true_positive, true_negative, false_positive, false_negative)  



#### Evaluate psi-blast

In [ ]:
conf_matrix_psiblast, metric_results_psiblast = metrics_results(positions_psiblast, set_protein_containing_domain, num_sequences )
print('CLASSIFICATION METRICS')
for k in metric_results_psiblast:
    print(f'{k}: {metric_results_psiblast[k]:.3}')
    
print('\nCONFUSION MATRIX')    
conf_matrix_psiblast

CLASSIFICATION METRICS
Accuracy: 0.999
Weighted accuracy: 1.0
Precision: 0.223
Sensitivity: 1.0
Specificity: 0.999
MCC: 0.472
F-score: 0.364

CONFUSION MATRIX


Actual positive  Actual negative
Positive pred              150              524
Negative pred                0           563298

#### Evaluate hmm

In [ ]:
conf_matrix_hmm, metric_results_hmm = metrics_results(positions_hmm, set_protein_containing_domain, num_sequences )
print('CLASSIFICATION METRICS')
for k in metric_results_hmm:
    print(f'{k}: {metric_results_hmm[k]:.3}')
    
print('\nCONFUSION MATRIX')    
conf_matrix_hmm

CLASSIFICATION METRICS
Accuracy: 1.0
Weighted accuracy: 0.993
Precision: 0.359
Sensitivity: 0.987
Specificity: 1.0
MCC: 0.595
F-score: 0.527

CONFUSION MATRIX


Actual positive  Actual negative
Positive pred              148              264
Negative pred                2           563558

### Comment on the results

We have so many false positives and very few false negatives. Probably we should try to retrieve less hits, since their correct number is 150, but our models selected much more sequences (674 with psiblast and 412 with hmm).

## Ex 9

#### Read the ground truth file for the positions

In [ ]:
path = datapath + 'protein-matching-PF08543.json'

with open(path, "r") as file:
    position_file = json.load(file)
    
ground_truth_positions = {}
for elem in position_file:
    id = elem['metadata']['accession']
    length = elem['metadata']['length']

    ground_truth_positions.setdefault(id, dict()).update([('length', length), ('positions', set())])

    for entry in elem['entries']:
        if entry['accession']=='PF08543':
            for location in entry['entry_protein_locations']:

                position_tuple=(str(location['fragments'][0]['start']),
                                str(location['fragments'][0]['end']))

                ground_truth_positions[id]['positions'].add(position_tuple)

In [ ]:
print(len(ground_truth_positions))

150


In [ ]:
for item in ground_truth_positions.items():
  print(item)

('A1ADT5', {'length': 283, 'positions': {('88', '265')}})
('A3N2D3', {'length': 286, 'positions': {('64', '256')}})
('A5UA83', {'length': 288, 'positions': {('71', '264')}})
('A5WB73', {'length': 290, 'positions': {('74', '257')}})
('A6VEZ4', {'length': 288, 'positions': {('77', '259')}})
('A6VNE5', {'length': 286, 'positions': {('62', '259')}})
('A7N5Q6', {'length': 289, 'positions': {('75', '260')}})
('A7ZPL9', {'length': 283, 'positions': {('88', '265')}})
('A8A2R4', {'length': 283, 'positions': {('88', '265')}})
('B0KR83', {'length': 290, 'positions': {('76', '257')}})
('B0UUD2', {'length': 286, 'positions': {('64', '260')}})
('B1IX53', {'length': 283, 'positions': {('88', '265')}})
('B1JFM7', {'length': 290, 'positions': {('76', '257')}})
('B1LML3', {'length': 283, 'positions': {('88', '265')}})
('B1XA89', {'length': 283, 'positions': {('88', '265')}})
('B2JCI0', {'length': 287, 'positions': {('74', '261')}})
('B2TX08', {'length': 283, 'positions': {('88', '265')}})
('B3H2H2', {'l

#### Useful functions to compute metrics

CONFIDENTIAL!!! :-)

In $\verb!utils.py!$ there are 5 functions:

- $\verb!computeMetrics!$
- $\verb!createPositionSet!$
- $\verb!evaluateSequencesSH2!$
- $\verb!getSeqLength!$
- $\verb!evaluatePositionsSH2!$

But for point 9) are needed only:
- $\verb!computeMetrics!$
- $\verb!createPositionSet!$
- $\verb!getSeqLength!$
- $\verb!evaluatePositionsSH2!$

So I use only them in the following.

In [ ]:
def createPositionSet(position_set):
    """
    position_set = {('88', '265'), ...}
    return a set {88, 89, ..., 265, ...}
    """
    positions = set()
    for pos in position_set: # pos is a tuple of two strings
        p = [i for i in range( int(pos[0]), int(pos[1])+1 )]
        positions = positions.union(p)
    return positions

This is the original function.

I'm not able to run it because it calls a command line in Linux. 

If you can, please try to see if the below function provides the same result as my version.

I get the following error...

`Error: [blastdbcmd] DB contains no accession info.`

In [ ]:
############### ORIGINAL FUNCTION
'''
def getSeqLength(entry):
    """
    Get sequence length from reference dataset
    """   
    cmd = 'blastdbcmd -entry "{}" -db ../../data/SwissProt_humans_reference_all.fasta'.format(entry)
    results = subprocess.run(
            cmd, shell=True, universal_newlines=True, 
            stdout=subprocess.PIPE
            ).stdout
    results = results.split('\n')
    results.pop(0)
    return len("".join(results))
'''

In [ ]:
path = datapath + 'uniprot_sprot.fasta'
all_sequences = SeqIO.parse(path, 'fasta')

In [ ]:
##################### MY VERSION
def getSeqLength(entry, all_sequences):

  """
  Get sequence length from reference dataset
  """ 
#   path = datapath + 'uniprot_sprot.fasta'
#   all_sequences = SeqIO.parse(path, 'fasta')

  list_all_sequences = []
  for sequence in all_sequences:
      name = sequence.id # name is in the form sp|Q6GZX4|001R_FRG3G
      name = name.split('|')[1]
      if name==entry:
        #print(sequence.seq)
        #print(len(sequence.seq))
        length=len(sequence.seq)
  return length


In [ ]:
#-------------------------DEBUG
entry='Q6GZX4'
getSeqLength(entry, all_sequences)

256

In [ ]:
def evaluatePositions(predicted_positions, ground_truth_positions):
    """
    predicted_positions: {'A0A0H2WZY4': {('162', '278')}, ...}
    ground_truth_positions: {'A1ADT5': {'length': 283, 'positions': {('88', '265')}}, ...}
    
    """
    global_tp, global_fp, global_fn, global_tn = [], [], [], []
    
    # scores per sequence
    log_scores = []

    for id in predicted_positions:
        if id in ground_truth_positions:
            sequence_length = ground_truth_positions[id]['length']
            ground_truth_positions_seq = createPositionSet(ground_truth_positions[id]['positions']) 
            ####################################################################################
            # Here there was an error: there was reference_positions = createPositionSet(reference_positions[seqid]['positions']), 
            # which modifies reference_positions! So we had correct results only for the first protein!
            ####################################################################################
            predicted_positions_seq = createPositionSet(predicted_positions[id])    

            overlapping_positions  = ground_truth_positions_seq & predicted_positions_seq

            true_positive = len(overlapping_positions)
            false_positive = len(predicted_positions_seq - overlapping_positions)
            false_negative = len(ground_truth_positions_seq -  overlapping_positions)
            true_negative  = sequence_length - true_positive - false_positive - false_negative

            global_tp.append(true_positive)
            global_fp.append(false_positive)
            global_fn.append(false_negative)
            global_tn.append(true_negative)
            
            log_scores.append(metrics(true_positive, true_negative, false_positive, false_negative))


        ##############
        # HA SENSO? ci mette una vita per calcolare la lunghezza delle sequence con getSeqLength 
        # e non so se sia utile/sensato comprendere le sequenze senza dominio in questo esercizio
        ##############
        # else:
        #     # false positive sequence
        #     predicted_positions_seq = createPositionSet(predicted_positions[id])

        #     true_positive = 0
        #     false_positive = len(predicted_positions_seq)
        #     false_negative = 0
        #     true_negative = 0
            
        #     lenSeq = getSeqLength(id)
            
        #     log_scores.append(
        #         {
        #             'Precision': 0,
        #             'Sensitivity': 0,
        #             'Specificity': lenSeq / (lenSeq + false_positive),
        #             'F1-score': 0,
        #             'MCC': 0
        #         }
        #     )

        #     list_tp.append(true_positive)
        #     list_fp.append(false_positive)
        #     list_fn.append(false_negative)
        #     list_tn.append(true_negative)
                   
    metrics_global = metrics(true_positive=sum(global_tp), true_negative=sum(global_tn), 
                             false_positive=sum(global_fp), false_negative=sum(global_fn))
    
    conf_matrix = pd.DataFrame(np.array([[sum(global_tp), sum(global_fp)], [sum(global_fn), sum(global_tn)]]), 
                               index=['Positive pred', 'Negative pred'], 
                               columns=['Actual positive', 'Actual negative'])
    
    return metrics_global, log_scores, conf_matrix

#### Evaluate psi-blast

In [ ]:
metrics_global_psiblast, log_score_psiblast, conf_matrix_psiblast_pos = evaluatePositions(positions_psiblast, ground_truth_positions)

In [ ]:
print('CLASSIFICATION METRICS')
for k in metrics_global_psiblast:
    print(f'{k}: {metrics_global_psiblast[k]:.3}')
print('\nCONFUSION MATRIX')    
conf_matrix_psiblast_pos

CLASSIFICATION METRICS
Accuracy: 0.795
Weighted accuracy: 0.715
Precision: 0.759
Sensitivity: 0.999
Specificity: 0.43
MCC: 0.569
F-score: 0.862

CONFUSION MATRIX


Actual positive  Actual negative
Positive pred            29573             9413
Negative pred               26             7099

In [ ]:
# This is the output considering also the retrieved sequences without the domain
metrics_global_psiblast

{'Accuracy': 0.2367829747669103,
 'F1-score': 0.3334987313222441,
 'MCC': 0.10470260431297877,
 'Precision': 0.20015431367638797,
 'Sensitivity': 0.9991215919456738,
 'Specificity': 0.05666642719733071}

In [ ]:
# old version
metrics_global_psiblast

{'Accuracy': 0.0033580173508824087,
 'F1-score': 0.003297408696239738,
 'MCC': 0.0016815567089388617,
 'Precision': 0.0016514270630994038,
 'Sensitivity': 1.0,
 'Specificity': 0.0017122360584731998}

In [ ]:
log_score_psiblast

[{'Accuracy': 0.9521072796934866,
  'F-score': 0.9512670565302145,
  'MCC': 0.9085663242927233,
  'Precision': 0.9070631970260223,
  'Sensitivity': 1.0,
  'Specificity': 0.9100719424460432,
  'Weighted accuracy': 0.9550359712230216},
 {'Accuracy': 0.9582577132486388,
  'F-score': 0.957169459962756,
  'MCC': 0.9198108458946662,
  'Precision': 0.9178571428571428,
  'Sensitivity': 1.0,
  'Specificity': 0.9217687074829932,
  'Weighted accuracy': 0.9608843537414966},
 {'Accuracy': 0.9398496240601504,
  'F-score': 0.96875,
  'MCC': 0.32307445639837323,
  'Precision': 0.9393939393939394,
  'Sensitivity': 1.0,
  'Specificity': 0.1111111111111111,
  'Weighted accuracy': 0.5555555555555556},
 {'Accuracy': 0.9714285714285714,
  'F-score': 0.9722222222222222,
  'MCC': 0.944400281603035,
  'Precision': 0.9459459459459459,
  'Sensitivity': 1.0,
  'Specificity': 0.9428571428571428,
  'Weighted accuracy': 0.9714285714285714},
 {'Accuracy': 0.9511278195488722,
  'F-score': 0.9744597249508841,
  'MCC': 

In [ ]:
# old version
score_sequence_psiblast

[{'Accuracy': 0.9521072796934866,
  'F1-score': 0.9512670565302145,
  'MCC': 0.9085663242927233,
  'Precision': 0.9070631970260223,
  'Sensitivity': 1.0,
  'Specificity': 0.9100719424460432},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.6630565583634176},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5018867924528302},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.6542056074766355},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5047438330170778},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.661188369152971},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5092250922509225},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5092250922509225},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Spec

#### Evaluate hmm

In [ ]:
metrics_global_hmm, log_score_hmm, conf_matrix_hmm_pos = evaluatePositions(positions_hmm, ground_truth_positions)

In [ ]:
print('CLASSIFICATION METRICS')
for k in metrics_global_hmm:
    print(f'{k}: {metrics_global_hmm[k]:.3}')
print('\nCONFUSION MATRIX')    
conf_matrix_hmm_pos

CLASSIFICATION METRICS
Accuracy: 0.894
Weighted accuracy: 0.904
Precision: 0.964
Sensitivity: 0.867
Specificity: 0.942
MCC: 0.785
F-score: 0.913

CONFUSION MATRIX


Actual positive  Actual negative
Positive pred            25323              951
Negative pred             3886            15377

In [ ]:
metrics_global_hmm

{'Accuracy': 0.00891588785046729,
 'F1-score': 0.00900850387814223,
 'MCC': -0.012017725454758646,
 'Precision': 0.004524971836274878,
 'Sensitivity': 0.9836734693877551,
 'Specificity': 0.004431508778518449}

In [ ]:
log_score_hmm

[{'Accuracy': 0.9734693877551021,
  'F-score': 0.9737373737373737,
  'MCC': 0.9471360338761622,
  'Precision': 0.964,
  'Sensitivity': 0.9836734693877551,
  'Specificity': 0.963265306122449,
  'Weighted accuracy': 0.9734693877551021},
 {'Accuracy': 0.9726277372262774,
  'F-score': 0.9696969696969697,
  'MCC': 0.9449035066900258,
  'Precision': 0.96,
  'Sensitivity': 0.9795918367346939,
  'Specificity': 0.966996699669967,
  'Weighted accuracy': 0.9732942682023304},
 {'Accuracy': 0.9398496240601504,
  'F-score': 0.967871485943775,
  'MCC': 0.5017285475203439,
  'Precision': 0.9601593625498008,
  'Sensitivity': 0.9757085020242915,
  'Specificity': 0.47368421052631576,
  'Weighted accuracy': 0.7246963562753036},
 {'Accuracy': 0.9516728624535316,
  'F-score': 0.9741550695825051,
  'MCC': 0.6104374621209737,
  'Precision': 0.9645669291338582,
  'Sensitivity': 0.9839357429718876,
  'Specificity': 0.55,
  'Weighted accuracy': 0.7669678714859438},
 {'Accuracy': 0.9712643678160919,
  'F-score': 

In [ ]:
score_sequence_hmm

[{'Accuracy': 0.9734693877551021,
  'F1-score': 0.9737373737373737,
  'MCC': 0.9471360338761622,
  'Precision': 0.964,
  'Sensitivity': 0.9836734693877551,
  'Specificity': 0.963265306122449},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.6867167919799498},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5145067698259188},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.51434034416826},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.6752910737386805},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.516504854368932},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5213178294573644},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Specificity': 0.5155038759689923},
 {'F1-score': 0,
  'MCC': 0,
  'Precision': 0,
  'Sensitivity': 0,
  'Speci